Documentation/proxy.md

```diff
@@ -6,9 +6,11 @@ etcd currently supports two proxy modes: `readwrite` and `readonly`. The default
 
 The proxy will shuffle the list of cluster members periodically to avoid sending all connections to a single member.
 
-The member list used by proxy consists of all client URLs advertised within the cluster, as specified in each members' `-advertise-client-urls` flag. If this flag is set incorrectly, requests sent to the proxy are forwarded to wrong addresses and then fail. Including URLs in the `-advertise-client-urls` flag that point to the proxy itself, e.g. http://localhost:2379, is even more problematic as it will cause loops, because the proxy keeps trying to forward requests to itself until its resources (memory, file descriptors) are eventually depleted. The fix for this problem is to restart etcd member with correct `-advertise-client-urls` flag. After client URLs list in proxy is recalculated, which happens every 30 seconds, requests will be forwarded correctly.
+The member list used by proxy consists of all client URLs advertised within the cluster, as specified in each member's `advertise-client-urls` flag that represents client URLs to advertise to the public.
 
-You should start the cluster machines first and then start the proxy as the proxy gets its list from reachable machines. If the proxy starts and is unable to reach a machine it will refresh aggressively. Once the machine list is non-empty the proxy will refresh at the original duration.
+etcd proxy first tries to get its endpoints(peerURLs) from `discovery` or `discovery-srv` flag, otherwise from `initial-cluster` flag. And then gets clientURLs from the first reachable member from the peerURLs. And this clientURLs is set by the flag `advertise-client-urls`. And proxy connects to this clientURLs every 30-second in order to refresh its endpoints. This is why we need both `initial-cluster` and `advertise-client-urls` flags for every member in cluster, and proxy does not need `advertise-client-urls` flag. If these flags aren't set correctly, requests to proxy would be forwarded to wrong addresses. Having URLs in `advertise-client-urls` that points to the proxy itself can cause proxy loops forwarding requests to itself until resource depletion. In this case, you need to restart etcd member with correct `advertise-client-urls` flag.
```

Comment:
An etcd proxy examines several command-line options to discover its endpoints (peer URLs). In order of precedence, these options are `discovery`, `discovery-srv`, and `initial-cluster`. Each option takes one or more etcd peer URLs as arguments.

After establishing a list of peer URLs in this manner, the proxy retrieves the list of client URLs from the first reachable peer. These client URLs are specified by the `advertise-client-urls` option to etcd, which takes one or more etcd client service endpoint URLs as arguments. The proxy then connects to the client URLs every thirty seconds to refresh this list of endpoints.

While etcd proxies therefore do not need to be given the `advertise-client-urls` option, as they retrieve it from the cluster, this implies that both `initial-cluster` and `advertise-client-urls` options must be set correctly for every non-proxy, first-order cluster member. Otherwise, requests to any etcd proxy would be forwarded improperly. Take special care not to give URL arguments to the `advertise-client-urls` option that point to the proxy itself, as such a configuration will cause the proxy to enter a suicidal loop, forwarding requests to itself until resources are exhausted. To correct either case, stop etcd and restart it with the correct URLs.